### PR TITLE
Handle ignore_above in synthetic source for flattened fields

### PR DESCRIPTION
**Pull request by lkts**
This PR handles synthetic source for `flattened` fields when `ignore_above` is used. Previously child fields longer than `ignore_above` were missing from synthetic source.

This PR uses a rather "brute force" approach to the problem. If there is any ignored child fields encountered during parsing we store _the entire object_ in `_ignored_source` field. There is a possible optimization to store only ignored child fields in dedicated stored fields and then merge such values with valid fields read from doc values. Optimized approach is more complex and requires restructuring more code so i propose to handle it as a separate issue.

Contributes to #106485.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java [hunk 1]
```diff
@@ -32,44 +36,64 @@ class FlattenedFieldParser {
...
     }
 
-    public List<IndexableField> parse(final DocumentParserContext context) throws IOException {
-        XContentParser parser = context.parser();
+    public List<IndexableField> parse(final DocumentParserContext documentParserContext) throws IOException {
+        XContentParser parser = documentParserContext.parser();
         XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
 
+        XContentBuilder rawDataForSyntheticSource = null;
+        if (documentParserContext.mappingLookup().isSourceSynthetic() && ignoreAbove < Integer.MAX_VALUE) {
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java:68]: Use `documentParserContext.canAddIgnoredField`.

File: server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java [hunk 2]
```diff
@@ -32,44 +36,64 @@ class FlattenedFieldParser {
...
+            var copyingParser = new CopyingXContentParser(parser);
+            rawDataForSyntheticSource = copyingParser.getBuilder();
+            parser = copyingParser;
+        }
+
         ContentPath path = new ContentPath();
         List<IndexableField> fields = new ArrayList<>();
 
+        var context = new Context(parser, documentParserContext);
         parseObject(context, path, fields);
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java:78]: I was thinking of having `parseObject` and children return boolean, to indicate if they encountered ignoredAbove. The context works too.

**Review by kkrik-es — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java [hunk 3]
```diff
@@ -32,44 +36,64 @@ class FlattenedFieldParser {
...
+        if (rawDataForSyntheticSource != null && context.isIgnoredValueEncountered()) {
+            // One or more inner fields are ignored due to `ignore_above`.
+            // Because of that we will store whole object as is in order to generate synthetic source.
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java:81]: Let's create an issue to track improving this which we can prioritise later.

**Review by salvatore-campagna — APPROVED** [review]

**Comment by lkts** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java (hunk 2 above)
Comment (lkts) [reply server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java:78]: That was my initial idea too but i am not a fan of unnamed booleans.

File: server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java (hunk 3 above)
Comment (lkts) [reply server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java:81]: #110239

**Pull request merged**